### ISSUE TITLE
System cursors switching

### ISSUE THREAD

**Issue by PapyChacal** [panda3d/panda3d#1049]
## Description

  I'd like to see Panda3D offer an interface to switch between system cursors, like the usual arrow cursor, the text cursor, or the finger cursor, for example.

## Use Case
  I'm working on a dear ImGui backend for Panda3D, and it would allow consistent cursor behavior in the dear ImGui interface while using only the Panda3D's interface. It would help for any Panda3D-based tool's GUI, I guess.
  Dear ImGui is great for such GUIs, that's why I'm working on a backend for Panda3D, hoping for integration on the Panda3D's side later, which would be easy once the backend is ready. If this is not considered useful, integrating dear ImGui in any Panda3D project will still be easier.

**Comment by rdb** [comment]
I would suggest that we amend WindowProperties to take a new "cursor type" enumeration property, which would contain a "default" value, a "hidden" value (replacing the hidden cursor property), and a list of built-in cursors such as the one [here](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor).

**Comment by PapyChacal** [comment]
I'm working on this, I can switch between two cursor types on X11 for the moment. I have questions:
- What about the "cursor filename" property: should it be used with a given cursor type, as hidden, or should it override the cursor type ?
- I'm new to the overall Panda3D's architecture: I found usages of cursor_filename in WinGraphicsWindows and X11GraphicsWindows only: are those the only places where I should implement this new property, or am I missing some?

**Comment by rdb** [comment]
Thanks for working on this!

Good question.  We could make a cursor type we call "custom" which, when set, uses the specified cursor_filename; or we could just say that the cursor_filename always takes precedence and Panda falls back to the specified cursor type if the cursor file can't be loaded.  I am at the moment leaning towards the latter option.

The cursor_filename is defined in windowProperties.h and implemented in the various *GraphicsWindow.h classes.  Ideally we'd also add it to cocoaGraphicsWindow.h for macOS but perhaps someone else can add it if you are unable to.

**Comment by PapyChacal** [comment]
My pleasure :slightly_smiling_face: 

I agree with the latter idea about cursor_filename and will work on this.

I don't have a macOS environment to properly check what I implement in there. I'll write something first and let someone test it!

**Comment by PapyChacal** [comment]
While working on this, I've seen that making these choices lead to a complication or inconsistency.
If we replace the cursor_hidden property by a cursor_type value **and** give total precedence to cursor_filename of cursor_type, now cursor_filename has precedence over the hidden cursor_type, which is not the current behavior if I understood well. (For the moment, cursor_hidden hide the cursor, whatever icon or system cursor it is, right ?)

I can mix both approaches and give precedence to cursor_filename over cursor_type, except if its value is "hidden". But this leads to logically complicated and less readable code: I think maybe having cursor_type as a fallback to cursor_filename and keep a separate cursor_hidden property would be more intuitive?

**Comment by rdb** [comment]
Yes, right now, the hidden cursor flag takes precedence over the cursor filename.

Maybe it's simpler to just make cursor_type take precedence over cursor_filename?

**Comment by PapyChacal** [comment]
So, precedence for cursor_type with a value indicating that the filename should be used?
Any way is fine for me, but I don't know which is the best for current users, as I'm completely new to this community.

Also, should I keep the set/get_cursor_hidden functions, as convenience functions to set or check the cursor_type? This would prevent the inconvenience of changing those calls for users.
